### PR TITLE
mzbuild: prepare for export to fuzz testing platform

### PR DESCRIPTION
**Pull request by benesch**
Prepare mzbuild for exporting to our new fuzz testing platform (codename
"ATP"):

  * Tweak the directory pre-image to be a more general-purpose copy
    pre-image plugin. This makes it possible for images to refer to
    files outside of their build context.

  * Add a new driver image which specifies how to orchestrate the other
    images.

  * Add a Python script which builds and pushes the necessary images to
    ATP.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6143)

### CODE REVIEW

File: test/atp/testdrive/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+MZFROM testdrive
+
+COPY tests /tests
```
Comment (philip-stoev) [inline test/atp/testdrive/Dockerfile:12]: I think the copy target needs to be ```/workdir``` .

File: test/atp/testdrive/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Copyright Materialize, Inc. All rights reserved.
```
Comment (philip-stoev) [inline test/atp/testdrive/Dockerfile:1]: You may wish to remove ```test/testdrive/Dockerfile``` from the repository then.

**Review by philip-stoev — APPROVED** [review]
Thank you!

File: test/atp/testdrive/Dockerfile (hunk 1 above)
Comment (benesch) [reply test/atp/testdrive/Dockerfile:12]: Ah good catch, thanks!

File: test/atp/testdrive/Dockerfile (hunk 2 above)
Comment (benesch) [reply test/atp/testdrive/Dockerfile:1]: Righto! Fixed. Thanks.

File: misc/python/materialize/mzbuild.py [hunk 3]
```diff
@@ -160,21 +160,38 @@ def extra(self) -> str:
...
+        if self.destination is None:
+            raise ValueError("mzbuild config is missing 'destination' argument")
+
+        self.matching = config.pop("matching", "*")
+
+    def run(self) -> None:
+        super().run()
+        for src in self.inputs():
+            dst = self.path / self.destination / Path(src).relative_to(self.source)
+            dst.parent.mkdir(parents=True, exist_ok=True)
```
Comment (cirego) [inline misc/python/materialize/mzbuild.py:190]: Given that `super().run()` is running `git clean`, should we bail the `destination` directory exists?

Comment (benesch) [reply misc/python/materialize/mzbuild.py:190]: It's not quite that simple. If you're copying files `src/foo/a` and `src/foo/b` then we'll call mkdir on `src/foo` twice. We could definitely try to be more clever, but not sure it's worth it.

Comment (cirego) [reply misc/python/materialize/mzbuild.py:190]: Right, because this is the parent directory. Makes sense to me, thanks!

**Pull request merged**